### PR TITLE
Rename multiplexer to varianter

### PR DESCRIPTION
**Pull request by ldoktor**
This PR renames avocado.core.multiplexer to avocado.core.varianter and then separates the mux-specific parts and moves them into newly created avocado.core.mux.

This change is not complete, but to be able to separate everything we need to first make the varianter to support plugins, which will be treated next.

v1: https://github.com/avocado-framework/avocado/pull/1632

Changes:

```
v2: Fixed (hopefully) all references to multiplexer/mux where it changed meaning
v2: Added missing `root` docstring
v2: Separated a bugfix into an extra patch
v2: Use `__class__` in `MuxTreeNode.__repr__`
```

### CODE REVIEW

**Comment by clebergnu** [comment]
@ldoktor this PR is failing on Travis.  Is it ready for review?

**Comment by ldoktor** [comment]
Hello @clebergnu, sorry about that, it was passing, then I remember I need to fix just one thing and this one small adjustment broke the selftests. Anyway yes, it's ready for review...

**Comment by clebergnu** [comment]
@ldoktor OK, working on it.

File: avocado/core/job.py [hunk 1]
```diff
@@ -38,7 +38,7 @@
 from . import exceptions
 from . import job_id
 from . import output
-from . import multiplexer
+from . import varianter
```
Comment (ldoktor) [inline avocado/core/job.py]: Oups, true

File: avocado/core/jobdata.py [hunk 2]
```diff
@@ -30,7 +30,7 @@
 CONFIG_FILENAME = 'config'
 TEST_REFERENCES_FILENAME = 'test_references'
 TEST_REFERENCES_FILENAME_LEGACY = 'urls'
-MUX_FILENAME = 'multiplex'
+VARIANTS_FILENAME = 'multiplex'
```
Comment (ldoktor) [inline avocado/core/jobdata.py]: Right now nothing. I think we should change it when we come up with the definite solution of the multiplex jobdata storage (I mean the implementation not based on pickle)

File: avocado/core/runner.py [hunk 3]
```diff
@@ -481,8 +481,8 @@ def _iter_variants(template, mux):
         for variant, params in mux.itertests():
```
Comment (ldoktor) [inline avocado/core/runner.py]: Yep, good catch...

File: docs/source/Mux.rst [hunk 4]
```diff
@@ -1,10 +1,10 @@
-.. _mux:
+.. _variants:
 
-===================
-Test variants - Mux
-===================
+========================
+Test variants - Variants
```
Comment (ldoktor) [inline docs/source/Mux.rst]: OK, I'll use `Test Variants`.

File: docs/source/Mux.rst [hunk 5]
```diff
@@ -19,63 +19,63 @@ in reality it follows the way people are used to define dependencies,
...
 
-The ``Mux`` is a core part of avocado and one can see it as a ``multiplexed``
+The ``Varianter`` is a core part of avocado and one can see it as a ``multiplexed``
 database, which contains key/value pairs associated to given paths and
 as we are talking about a tree of those, we call the paths ``Nodes``.
 
-Mux allows iterating through all possible combinations which are stored in
-the database, which is called ``multiplexation``. Mux yields ``variants``,
+Varianter allows iterating through all possible combinations which are stored in
+the database, which is called ``multiplexation``. Varianter yields ``variants``,
```
Comment (ldoktor) [inline docs/source/Mux.rst]: Well currently only `multiplexer` is available in `varianter`. We define i as `multiplexed` database. Then we say it allows iterating throught combinations called `multiplexation`, which produces `variants`.

I think it's actually consistent with the current features and when we allow different varianter plugins, we have to rewrite this from scratch as the difference is too big. Do you really think it's worth trying to fix this version (which contains several other issues already)?

File: docs/source/Mux.rst [hunk 6]
```diff
@@ -19,63 +19,63 @@ in reality it follows the way people are used to define dependencies,
...
+Varianter API
+-------------
 
 .. warning:: This API is internal, we might change it at any moment. On the
              other hand we maintain ``avocado-virt`` plugin which uses this
              API so in such case we'd provide a patch there demonstrating
              the necessary changes.
 
-The ``Mux`` object is defined in ``avocado/core/multiplexer.py``, is always
+The ``Varianter`` object is defined in ``avocado/core/multiplexer.py``, is always
```
Comment (ldoktor) [inline docs/source/Mux.rst]: Sorry, I'll fix that

File: docs/source/Mux.rst [hunk 7]
```diff
@@ -85,15 +85,15 @@ Given these you should be able to implement any kind of parser or params
...
+``Varianter`` works: `yaml_to_mux plugin`_
 
 
 Yaml_to_mux plugin
 ==================
 
 In order to get a good coverage one always needs to execute the same test
 with different parameters or in various environments. Avocado uses the
-term ``Multiplexation`` or ``Mux`` to generate multiple variants of the same
+term ``Multiplexation`` or ``Variants`` to generate multiple variants of the same
```
Comment (ldoktor) [inline docs/source/Mux.rst]: Oups, this sentence does not make sense. It should be just the `Multiplexation`...

The biggest problem in this documentation is that it was written as `Multiplexer` documentation, then it was spit to `Multiplex` and `yaml_to_mux` documentation and now we are trying to rename the `Multiplex` to `Varianter`, which at this time only support `Multiplexer` as the only available and hardcoded generator.

I don't think it's actually worth trying to perfectly define this fixed version and we should spend more time creating a nice generic `Varianter` documentation with support for various plugins, `yaml_to_mux` being one of them.

File: selftests/functional/test_replay_basic.py [hunk 8]
```diff
@@ -164,17 +164,17 @@ def test_run_replay_remotefail(self):
         msg = "Currently we don't replay jobs in remote hosts."
         self.assertIn(msg, result.stderr)
 
-    def test_run_replay_status_and_mux(self):
+    def test_run_replay_status_and_variants(self):
         """
-        Runs a replay job with custom a mux and using '--replay-test-status'
+        Runs a replay job with custom a variants and using '--replay-test-status'
```
Comment (ldoktor) [inline selftests/functional/test_replay_basic.py]: I think there are even two: How about `Runs a replay job with custom variants using '--replay-test-status'`

File: avocado/core/job.py [hunk 9]
```diff
@@ -441,24 +441,24 @@ def pre_tests(self):
         self._result_events_dispatcher.map_method('pre_tests', self)
 
     def run_tests(self):
-        mux = getattr(self.args, "mux", None)
-        if mux is None:
-            mux = multiplexer.Mux()
-        if not mux.is_parsed():   # Mux not yet parsed, apply args
+        variant = getattr(self.args, "xxx_variants", None)
```
Comment (adereis) [inline avocado/core/job.py]: `xxx` or `XXX` is commonly used in programming for something that is ugly, dirty or badly implemented. See https://en.wikipedia.org/wiki/Comment_(computer_programming)#Tags

`x-`, is commonly used to prefix options which are experimental.

(and your case is none of the above, if I understood you correctly)

**Comment by adereis** [comment]
In one of your changelogs, you describe the architecture as: 

> [...] abstract API for variants generation, multiplexer which will be in-core implementation of
one of the variants generation plugin and yaml_to_mux plugin which uses
this mutliplexer implementation to create variants from a yaml file.

This is very confusing. There's an *abstract API*, *in-core implementation*, *variants generation plugins*, and then a `yaml_to_mux` (note the "mux" keyword) *plugin which uses this multiplexer implementation*.

Somewhere there's also a comment that says something like "generic Varianter [...] with support for various plugins, yaml_to_mux being one of them". A change like this and all the discussions around it should be using a consistent and very clear vocabulary.

**Comment by apahim** [comment]
@ldoktor, commits are in good shape. For the `xxx_` in the var name, I agree it should not be used in this case. Could you send a new version with the fixes already requested? Then I will take a final look and if nothing calls my attention, we can consider it ready to merge.

**Pull request closed without merging**